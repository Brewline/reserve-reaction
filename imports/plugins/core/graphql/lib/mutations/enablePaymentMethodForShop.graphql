enablePaymentMethodForShop($input: EnablePaymentMethodForShopInput!) {
  enablePaymentMethodForShop(input: $input) {
    paymentMethods {
      name
      displayName
      isEnabled
      pluginName
    }
  }
}
