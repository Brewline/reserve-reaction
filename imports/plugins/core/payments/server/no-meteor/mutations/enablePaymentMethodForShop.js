import ReactionError from "@reactioncommerce/reaction-error";
import SimpleSchema from "simpl-schema";
import { paymentMethods as allPaymentMethods } from "/imports/plugins/core/core/server/no-meteor/pluginRegistration";

const paramsSchema = new SimpleSchema({
  isEnabled: Boolean,
  paymentMethodName: String,
  shopId: String
});

/**
 * @method enablePaymentMethodForShop
 * @summary Enables (or disables) payment method for a given shop
 * @param {Object} context -  an object containing the per-request state
 * @param {Object} input - EnablePaymentMethodForShopInput
 * @param {String} input.isEnabled - Whether to enable or disable specified payment method
 * @param {String} input.paymentMethodName - The name of the payment method to enable or disable
 * @param {String} input.shopId - The id of the shop to enable payment method on
 * @param {String} [input.clientMutationId] - An optional string identifying the mutation call
 * @return {Promise<Array<Object>>} Array<PaymentMethod>
 */
export default async function enablePaymentMethodForShop(context, input = {}) {
  paramsSchema.validate(input, { ignore: [SimpleSchema.ErrorTypes.KEY_NOT_IN_SCHEMA] });
  const { Shops } = context.collections;
  const { isEnabled, paymentMethodName, shopId } = input;

  if (!context.userHasPermission(["owner", "admin"], shopId)) {
    throw new ReactionError("access-denied", "Access denied");
  }

  if (!allPaymentMethods[paymentMethodName]) {
    throw new ReactionError("not-found", "Requested payment method is invalid");
  }

  const shop = await context.queries.shopById(context, shopId);
  if (!shop) throw new ReactionError("not-found", "Shop not found");

  const methods = new Set(shop.availablePaymentMethods);

  if (isEnabled) {
    methods.add(paymentMethodName);
  } else {
    methods.delete(paymentMethodName);
  }

  await Shops.updateOne(
    { _id: shop._id },
    { $set: { availablePaymentMethods: Array.from(methods) } }
  );

  return context.queries.paymentMethods(context, shopId);
}
