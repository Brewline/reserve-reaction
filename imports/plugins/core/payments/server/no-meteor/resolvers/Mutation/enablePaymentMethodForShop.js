import { decodeShopOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/shop";

/**
 * @name "Mutation.enablePaymentMethodForShop"
 * @method
 * @memberof Payment/GraphQL
 * @summary resolver for the enablePaymentMethodForShop GraphQL mutation
 * @param {Object} parentResult - unused
 * @param {Object} args.input - EnablePaymentMethodForShopInput
 * @param {String} args.input.isEnabled - Whether to enable or disable specified payment method
 * @param {String} args.input.paymentMethodName - The name of the payment method to enable or disable
 * @param {String} args.input.shopId - The id of the shop to enable payment method on
 * @param {String} [args.input.clientMutationId] - An optional string identifying the mutation call
 * @param {Object} context - an object containing the per-request state
 * @return {Promise<Array<Object>>} EnablePaymentMethodForShopPayload
 */
export default async function enablePaymentMethodForShop(parentResult, { input }, context) {
  const { clientMutationId } = input;
  const paymentMethods = await context.mutations.enablePaymentMethodForShop(context, {
    ...input,
    shopId: decodeShopOpaqueId(input.shopId)
  });

  return {
    clientMutationId,
    paymentMethods
  };
}
