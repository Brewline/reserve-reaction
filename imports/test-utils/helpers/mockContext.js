const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  shopId: "FAKE_SHOP_ID",
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "Groups",
  "Inventory",
  "MediaRecords",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "Tags",
  "Templates",
  "Themes",
  "Translations",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
