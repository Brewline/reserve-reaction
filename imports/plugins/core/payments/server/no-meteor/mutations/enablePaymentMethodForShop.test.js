import Factory from "/imports/test-utils/helpers/factory";
import enablePaymentMethodForShop from "./enablePaymentMethodForShop";
import mockContext from "/imports/test-utils/helpers/mockContext";

jest.mock("/imports/plugins/core/core/server/no-meteor/pluginRegistration", () => ({
  paymentMethods: {
    mockPaymentMethod: {
      name: "mockPaymentMethod",
      displayName: "Mock!",
      pluginName: "mock-plugin"
    }
  }
}));

const fakeShop = Factory.Shop.makeOne();
const mockEnablePaymentMethod = jest.fn().mockName("enablePaymentMethodForShop");
const mockPaymentMethods = jest.fn().mockName("paymentMethods");
const mockShopById = jest.fn().mockName("shopById");

beforeAll(() => {
  mockContext.queries = {
    paymentMethods: mockPaymentMethods,
    shopById: mockShopById
  };
  mockContext.mutations = { enablePaymentMethodForShop: mockEnablePaymentMethod };
});

beforeEach(() => {
  jest.resetAllMocks();
  mockShopById.mockClear();
  mockEnablePaymentMethod.mockClear();
  fakeShop.availablePaymentMethods = [];
});

test("throws if userHasPermission returns false", async () => {
  mockContext.userHasPermission.mockReturnValue(false);
  mockShopById.mockReturnValue(fakeShop);

  await expect(enablePaymentMethodForShop(mockContext, mockContext.shopId)).rejects.toThrowErrorMatchingSnapshot();
});

test("errors on missing arguments", async () => {
  mockContext.userHasPermission.mockReturnValue(true);
  mockShopById.mockReturnValue(fakeShop);

  await expect(enablePaymentMethodForShop(mockContext, {})).rejects.toThrowErrorMatchingSnapshot();
});

test("errors on invalid payment method", async () => {
  mockContext.userHasPermission.mockReturnValue(true);
  mockShopById.mockReturnValue(fakeShop);

  await expect(enablePaymentMethodForShop(mockContext, {
    shopId: fakeShop._id,
    paymentMethodName: "does not exist",
    isEnabled: true
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("errors on invalid shop", async () => {
  mockContext.userHasPermission.mockReturnValue(true);
  mockShopById.mockReturnValue();

  await expect(enablePaymentMethodForShop(mockContext, {
    shopId: "does not exist",
    paymentMethodName: "mockPaymentMethod",
    isEnabled: true
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("enables payment method for valid shop", async () => {
  fakeShop.availablePaymentMethods = ["mockPaymentMethod"];
  mockContext.userHasPermission.mockReturnValue(true);
  mockShopById.mockReturnValue(fakeShop);
  mockPaymentMethods.mockReturnValue([{
    name: "mockPaymentMethod",
    displayName: "Mock!",
    pluginName: "mock-plugin",
    isEnabled: true
  }]);

  await expect(enablePaymentMethodForShop(mockContext, {
    shopId: fakeShop._id,
    paymentMethodName: "mockPaymentMethod",
    isEnabled: true
  })).resolves.toEqual([{
    name: "mockPaymentMethod",
    displayName: "Mock!",
    pluginName: "mock-plugin",
    isEnabled: true
  }]);

  expect(mockPaymentMethods).toHaveBeenCalledWith(mockContext, fakeShop._id);
});

test("disables payment method for valid shop", async () => {
  mockContext.userHasPermission.mockReturnValue(true);
  mockShopById.mockReturnValue(fakeShop);
  mockPaymentMethods.mockReturnValue([{
    name: "mockPaymentMethod",
    displayName: "Mock!",
    pluginName: "mock-plugin",
    isEnabled: false
  }]);

  await expect(enablePaymentMethodForShop(mockContext, {
    shopId: fakeShop._id,
    paymentMethodName: "mockPaymentMethod",
    isEnabled: false
  })).resolves.toEqual([{
    name: "mockPaymentMethod",
    displayName: "Mock!",
    pluginName: "mock-plugin",
    isEnabled: false
  }]);

  expect(mockPaymentMethods).toHaveBeenCalledWith(mockContext, fakeShop._id);
});
