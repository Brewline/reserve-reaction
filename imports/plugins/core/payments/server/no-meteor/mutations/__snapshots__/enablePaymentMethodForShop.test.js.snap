// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errors on invalid payment method 1`] = `"Requested payment method is invalid"`;

exports[`errors on invalid shop 1`] = `"Shop not found"`;

exports[`errors on missing arguments 1`] = `"Is enabled is required"`;

exports[`throws if userHasPermission returns false 1`] = `"The first argument of validate() must be an object"`;
